### ISSUE TITLE
Kubelet sometimes creates a second container when it shouldn't

### ISSUE THREAD

**Issue by ncdc** [kubernetes/kubernetes#23636]
If you create a pod that will never be able to run successfully, such as this:

``` yaml
apiVersion: v1
kind: Pod
metadata:
  name: bad
spec:
  containers:
  - name: bad
    image: busybox
    command:
    - /blarg
  restartPolicy: Never
```

sometimes you'll see 1 container that is created and then fails to start, and then the pod is marked as failed. Other times you'll see that the Kubelet creates 2 containers before marking the pod as failed. The following are kubelet log line numbers that show the sequence of events:

```
7577 - syncPod (1st time - for new pod)
7637 - PLEG starts relisting
7662 - 1st container created for pod
7663 - PLEG inspects container. Container isn't running yet
7666 - PLEG records "unknown" status
7672 - error running container
7690 - PLEG starts relisting again
7691 - syncPod (2nd time, status cache still has "unknown")
7714 - 2nd container created for pod
7731 - PLEG inspects containers and sees 2 dead containers now
7733 - PLEG updates status to dead
```

It looks like the pod worker is passing a slightly out of date pod status to syncPod.

@kubernetes/rh-cluster-infra @kubernetes/sig-node

**Comment by yujuhong** [comment]
As discussed on slack, my theory was that pleg failed to detect that there was a change in the container state.
More specifically,
 1. container created. Container status: "created"
 2. PLEG relists all container/pods and update the cache.
 3. kubelet tried starting the container and failed. Container status: "created", exit code: non-zero
 4. PLEG relists all container/pods and update the cache.

In (4), PLEG saw the the container state remains at "created", and decided *not* to update the cache -- even though the container now has a non-zero exit code and should be considered an "exited" container by kubelet.

In short, the root cause is that there is no state transition between "created" and "created but failed to start" (from `docker ps -a`). We need to consider either
 - Adding more detection mechanism in PLEG for such changes
 - Patch SyncPod to treat "created" containers as dead containers.

**Comment by ncdc** [comment]
I added debug logging and @yujuhong's theory is correct. syncPod waits for the cache to be updated, the PLEG relists, and because the `plegContainerState` both before and after was `unknown`, the entry in the cache is not updated, and things proceed from there.

**Comment by aveshagarwal** [comment]
@yujuhong It seems that the first option might be better (if possible) based on exit codes. As I think the exit code for "created but not yet started" is always 0, and the code is nonzero for "created but failed to start". By dealing like this, it is possible that the unknown state can be handled in a better way too. 

It seems to me that "created" (unknown state) containers are anyway being treated as dead containers because restart policy is not honored for them and they are always restarted.

**Comment by Random-Liu** [comment]
Ref #22607. :)

**Comment by aveshagarwal** [comment]
@Random-Liu a side effect of 22607 (and possibly in combination with this issue), one could see restart count > 0 even when restart policy is never.

**Comment by ncdc** [comment]
@Random-Liu this isn't exactly 22607, but it is related. The issue here is that there isn't enough information in ContainerState (https://github.com/kubernetes/kubernetes/blob/79012f6d5329b22fd77000a111d76dd8796c3e30/pkg/kubelet/container/runtime.go#L205-L210) to distinguish between "created but not yet started" and "created and failed to start".

**Comment by Random-Liu** [comment]
@aveshagarwal @ncdc Thanks, I just find that some information in #22607 is useful here. :)

**Comment by yujuhong** [comment]
> @yujuhong It seems that the first option might be better (if possible) based on exit codes. As I think the exit code for "created but not yet started" is always 0, and the code is nonzero for "created but failed to start". By dealing like this, it is possible that the unknown state can be handled in a better way too.
> 
> It seems to me that "created" (unknown state) containers are anyway being treated as dead containers because restart policy is not honored for them and they are always restarted.

@aveshagarwal, pleg relies on `docker ps -a` to detect "changes". Once it detects a change, it'd further `docker inspect` the container to populate all the information. The problem you mentioned won't happen normally.

This is a special case because the container status from `docker ps -a` remains "Created" even after the startup failure. `docker ps` does not give us an extra information to distinguish the two states, so a further inspection is not triggered. Arguably, docker should've returned an exited status since there is an exit code... Also, the impact of this issue is limited IMO since all it causes is creating an extra container (but can't start it) when restart policy is never.

We can either
 1. Document this as a known issue or
 2. Modify GetPods() to do an extra `docker inspect` if the a container status is "created" (which may add some non-trivial overhead if many containers remain in the state forever, and PLEG relists every second).

I'm open to be convinced this is a more severe problem though. @ncdc @aveshagarwal, any thoughts?

**Comment by aveshagarwal** [comment]
@yujuhong I am not sure either if it is a more severe problem. I have only seen it with specially crafted incorrect container command. In that case, the first option would be better.

**Comment by aveshagarwal** [comment]
@yujuhong @ncdc Would like to know if we want to keep it open and take further action, or close by documenting this as know issue/working as expected?

**Comment by yujuhong** [comment]
@aveshagarwal, I don't think we will take further action for the 1.3 release since the impact is very limited. I would like to keep the issue open to see for a while to see if this happens with a newer docker version and whether we can alleviate the impact (which is already very limited) for the 1.4 release.

**Comment by liggitt** [comment]
there have been no updates to this issue in 20 months. is it still being observed? if not, can we close this?
